File: app/src/main/java/com/qbapps/claudeusage/notification/UsageThresholdEvaluator.kt
package com.qbapps.claudeusage.notification

/**
 * Determines whether usage crossed any alert thresholds on this refresh,
 * returning only the highest crossed threshold when applicable.
 */
object UsageThresholdEvaluator {
    val SESSION_THRESHOLDS = listOf(75, 80, 85, 90, 100)
    val WEEKLY_THRESHOLDS = listOf(70, 80, 90, 100)

    fun highestReachedThreshold(
        currentUtilization: Double?,
        thresholds: List<Int> = SESSION_THRESHOLDS
    ): Int? {
        val current = currentUtilization?.coerceIn(0.0, 100.0) ?: return null
        return thresholds.lastOrNull { threshold -> current >= threshold }
    }

    fun highestCrossedThreshold(
        previousUtilization: Double?,
        currentUtilization: Double?,
        thresholds: List<Int> = SESSION_THRESHOLDS
    ): Int? {
        val current = currentUtilization?.coerceIn(0.0, 100.0) ?: return null
        val currentHighest = highestReachedThreshold(currentUtilization, thresholds) ?: return null
        if (previousUtilization == null) return currentHighest

        val previous = previousUtilization.coerceIn(0.0, 100.0)
        return thresholds.lastOrNull { threshold ->
            previous < threshold && current >= threshold
        }
    }
}

File: ios/OpenUsageCore/Sources/OpenUsageCore/Notifications/UsageThresholdEvaluator.swift
import Foundation

/// Faithful port of `notification/UsageThresholdEvaluator.kt` — the Claude
/// 5-hour session milestone ladder.
public enum UsageThresholdEvaluator {
    public static let thresholds = [75, 80, 85, 90, 100]

    public static func highestReachedThreshold(currentUtilization: Double?) -> Int? {
        guard let currentUtilization else { return nil }
        let current = clamp(currentUtilization, 0.0, 100.0)
        return thresholds.last { current >= Double($0) }
    }

    public static func highestCrossedThreshold(
        previousUtilization: Double?,
        currentUtilization: Double?
    ) -> Int? {
        guard let currentUtilization else { return nil }
        let current = clamp(currentUtilization, 0.0, 100.0)
        guard let currentHighest = highestReachedThreshold(currentUtilization: currentUtilization) else {
            return nil
        }
        guard let previousUtilization else { return currentHighest }

        let previous = clamp(previousUtilization, 0.0, 100.0)
        return thresholds.last { threshold in
            previous < Double(threshold) && current >= Double(threshold)
        }
    }
}
